File: src/main/java/io/github/mschout/gitlab/toggltimer/timer/GitLabIssue.java
package io.github.mschout.gitlab.toggltimer.timer;

import lombok.AccessLevel;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import lombok.val;
import org.springframework.web.util.UriComponentsBuilder;

@RequiredArgsConstructor(access = AccessLevel.PRIVATE)
@Getter
class GitLabIssue {

  private final String groupName;

  private final String projectPath;

  private final Long issueNumber;

  static GitLabIssue fromUrl(String url) {
    var ucb = UriComponentsBuilder.fromUriString(url).build();

    // "group/project/-/issues/12345"
    var pathSegments = ucb.getPathSegments();
    if (pathSegments.size() < 5) {
      throw new IllegalArgumentException("Invalid GitLab issue URL: " + url);
    }

    val groupName = pathSegments.get(0);
    val projectPath = pathSegments.get(1);
    val issueNumber = Long.parseLong(pathSegments.get(4));

    return new GitLabIssue(groupName, projectPath, issueNumber);
  }
}

File: src/main/kotlin/io/github/mschout/gitlab/toggltimer/timer/GitLabIssue.kt
package io.github.mschout.gitlab.toggltimer.timer

import org.springframework.web.util.UriComponentsBuilder

data class GitLabIssue(val groupName: String, val projectPath: String, val issueNumber: Long) {
  companion object {
    fun fromUrl(url: String): GitLabIssue {
      val pathSegments = UriComponentsBuilder.fromUriString(url).build().pathSegments
      require(pathSegments.size >= 5) { "Invalid GitLab issue URL: $url" }
      return GitLabIssue(
          groupName = pathSegments[0],
          projectPath = pathSegments[1],
          issueNumber = pathSegments[4].toLong(),
      )
    }
  }
}
